# Problem Sets 1
* Name: Yeji Kim 
* Student ID: 1863270
* Collaborators: None

# 1. Base Python 

### 1.1 Computing

1. (1pt) How many seconds are there in year?

In [1]:
60 * 60 * 24* 365

31536000

2. (1pt) How many seconds is a typical human lifetime? Use the previously computed seconds in year.

In [2]:
31536000 * 85

2680560000

Explanation: I used human life time as 85 years. 

3. (1pt) What is your age in seconds?

In [3]:
60 * 60 * 24* 365 * (24 + 1/12)

759492000.0

Explanation: I was born in Dec.11.1998. So (24 + 1/12) is my current age.

4. (1pt) Create a logical variable that is true if you are more than 700M seconds old. Use logical
operators, not if/else!

In [4]:
age_in_sec = (60 * 60 * 24* 365 * (24 + 1/12)) > 700000000
age_in_sec

True

5. (3pt) Compute the following mathematical formulas:

In [5]:
0.4*(1-0.4)

0.24

In [6]:
(3/2) * (1 + (2.5/2))**(-3-1)

0.058527663465935076

In [7]:
(3 * (2**2)) / (((3-1)**2) * (3-2))

3.0

### 1.2 Lists

1. (1pt) Create a list ‘movies‘ that contains the names of at least six movies you like

In [8]:
movies = ["About Times", "Begin Again", "Intern", "Lala Land", "The Devil Wears Prada", "Mean Girls"]
movies

['About Times',
 'Begin Again',
 'Intern',
 'Lala Land',
 'The Devil Wears Prada',
 'Mean Girls']

2. (2pt) Using indexing and slicing, Create a list of three first movies in the list

In [9]:
first_three_movies = movies[:3]
first_three_movies

['About Times', 'Begin Again', 'Intern']

3. (2pt) Use slicing and list concatenation to create a list of the first two and the last two movies

In [10]:
first_two_last_two_movies = movies[:2] + movies[-2:]
first_two_last_two_movies

['About Times', 'Begin Again', 'The Devil Wears Prada', 'Mean Girls']

4. (3pt) Use slicing to list the movies in the opposite order

In [11]:
movies[::-1]

['Mean Girls',
 'The Devil Wears Prada',
 'Lala Land',
 'Intern',
 'Begin Again',
 'About Times']

### 1.3 Loops

1. (3pt) Create a list ‘numbers‘ that is the numbers 70 through 79 in the following manner: create an
empty list and add numbers to it in a loop (do not use list comprehension or ‘list‘ function here!)
Hint: Consult python notes, ch 2.4.1 Lists.

In [12]:
numbers = []
for i in range(70,80):
    numbers.append(i)
numbers   

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

2. (3pt) Use a loop to compute the mean value of the list: add the values to their sum in a loop, and
at the end divide the sum by the number of the values. 

In [13]:
sum = 0
for i in numbers:
    sum += i
mean = sum / len(numbers)
mean

74.5

3. (3pt) Use loop to compute sum of all integers from 1 till 100.

In [14]:
sum = 0
for i in range(1,101):
    sum += i
sum

5050

4. (3pt) Assign people to seats. Consider names Adam, Ashin, Inukai, Tanaka, and Ikki; and seats 33,
12, 45, 2 and 17. Print seat assignments by assigning each name to the corresponding seat.

In [15]:
names = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
seat = [33, 12, 45, 2, 17]
for i in range(5):
    print(names[i] + ": " + str(seat[i]))

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


### 1.4 Dicts and Sets

1. (10pt) Create a word frequency table using dicts that take a looong string (feel free to copy-paste
some text here), splits it into individual words, and counts the number of occurrencies of each word
(and prints the result). Let’s ignore punctuation. Your code should run over individual words and
increase the counter for that word, stored in a list. It has to check if a word exists in the dictionary,
and if not, take an appropriate action.

In [16]:
song = ("Santa, tell me if you are really there Do not make me fall in love again If he will not be here next year Santa, tell me if he really cares Cause I cannot give it all away If he will not be here next year Feeling Christmas all around And I am trying to play it cool But it is hard to focus when I see him walking cross the room Let it Snow is blasting out But I will not get in the mood I am avoiding every mistletoe until I know It is true love that he thinks of So next Christmas I am not all alone, boy")
song = song.lower().split()
song

['santa,',
 'tell',
 'me',
 'if',
 'you',
 'are',
 'really',
 'there',
 'do',
 'not',
 'make',
 'me',
 'fall',
 'in',
 'love',
 'again',
 'if',
 'he',
 'will',
 'not',
 'be',
 'here',
 'next',
 'year',
 'santa,',
 'tell',
 'me',
 'if',
 'he',
 'really',
 'cares',
 'cause',
 'i',
 'cannot',
 'give',
 'it',
 'all',
 'away',
 'if',
 'he',
 'will',
 'not',
 'be',
 'here',
 'next',
 'year',
 'feeling',
 'christmas',
 'all',
 'around',
 'and',
 'i',
 'am',
 'trying',
 'to',
 'play',
 'it',
 'cool',
 'but',
 'it',
 'is',
 'hard',
 'to',
 'focus',
 'when',
 'i',
 'see',
 'him',
 'walking',
 'cross',
 'the',
 'room',
 'let',
 'it',
 'snow',
 'is',
 'blasting',
 'out',
 'but',
 'i',
 'will',
 'not',
 'get',
 'in',
 'the',
 'mood',
 'i',
 'am',
 'avoiding',
 'every',
 'mistletoe',
 'until',
 'i',
 'know',
 'it',
 'is',
 'true',
 'love',
 'that',
 'he',
 'thinks',
 'of',
 'so',
 'next',
 'christmas',
 'i',
 'am',
 'not',
 'all',
 'alone,',
 'boy']

In [17]:
song_word_counts = {}
for i in song:
    if i in song_word_counts.keys():
        song_word_counts[i] += 1
    else:
        song_word_counts[i] = 1
song_word_counts  

{'santa,': 2,
 'tell': 2,
 'me': 3,
 'if': 4,
 'you': 1,
 'are': 1,
 'really': 2,
 'there': 1,
 'do': 1,
 'not': 5,
 'make': 1,
 'fall': 1,
 'in': 2,
 'love': 2,
 'again': 1,
 'he': 4,
 'will': 3,
 'be': 2,
 'here': 2,
 'next': 3,
 'year': 2,
 'cares': 1,
 'cause': 1,
 'i': 7,
 'cannot': 1,
 'give': 1,
 'it': 5,
 'all': 3,
 'away': 1,
 'feeling': 1,
 'christmas': 2,
 'around': 1,
 'and': 1,
 'am': 3,
 'trying': 1,
 'to': 2,
 'play': 1,
 'cool': 1,
 'but': 2,
 'is': 3,
 'hard': 1,
 'focus': 1,
 'when': 1,
 'see': 1,
 'him': 1,
 'walking': 1,
 'cross': 1,
 'the': 2,
 'room': 1,
 'let': 1,
 'snow': 1,
 'blasting': 1,
 'out': 1,
 'get': 1,
 'mood': 1,
 'avoiding': 1,
 'every': 1,
 'mistletoe': 1,
 'until': 1,
 'know': 1,
 'true': 1,
 'that': 1,
 'thinks': 1,
 'of': 1,
 'so': 1,
 'alone,': 1,
 'boy': 1}

2. (3pt) Use set to find how many different words you have in your text above.
Note: you can get this information from your word frequency dict too, but this task is about using
set.

In [18]:
len(set(song))

67

In [19]:
# check if above is correct
len(song_word_counts)

67

### 1.5 Comprehension

1. (3pt) Use list comprehension to create a list of squares of numbers 1..10 (i.e. 1, 4, 9, ..., 100)

In [20]:
[i**2 for i in range(1,11)]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

2. (3pt) Use dict comprehension to create a dict of numbers 1..10 and their squares (i.e. 1:1, 2:4, ...,
10:100).

In [21]:
{i:i**2 for i in range(1,11)}

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}

### 1.6 Functions and Data Transformations 

1. (8pt) Write a function that takes in time in the numeric form (i.e. not a string) of HHMM (hours-
minutes), and returns it in the numeric form of HH.HH (hours + fractions of hours). For instance,
1015 →10.25 (10 hrs 15 mins →10.25 hrs). Demonstrate it works using values 1015 and 345.
The function should return (not print) the result as a number.
Hint: use modulo operator % and integer division operator //. Modulo of 100 gives you minutes
and integer division by 100 gives you hours

In [22]:
def time_transform (HHMM):
    hr = HHMM // 100
    minutes = round((HHMM % 100) /60, 2)
    return (hr + minutes)

In [23]:
# test
time_transform(1015)

10.25

In [24]:
# test
time_transform(345)

3.75

# Basic Data Frames

### 2.1 Setup

1. (2pt) Load the data

In [25]:
import pandas as pd

In [26]:
nycflight = pd.read_csv("data/nycflights13.csv.bz2", sep = '\t') 

2. (3 ×1pt) Do basic checks:
* How many variables (columns) is there in the data? Ensure you know the variables in the
data. Keep the documentation nearby.
* How many rows of data is there?
* print the first few lines of data. Does it look reasonable?

In [27]:
nycflight.shape

(336776, 19)

In [28]:
nycflight.head(5)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
2      1089     5      40  2013-01-01T10:00:00Z  
3      1576     5      45  2013-01-01T10:00:00Z  
4       762     6       0  2013-01-01T11:00:00Z

* The dataset has 19 columns. 
* The dataset has 336776 rows. 
* It seems reasonable because it has 19 columns and all rows and columns has its data. 

### 2.2 Basic Data Exploration

1. (5pt) How many NYC airports are included in this data? Which airports are these?

In [29]:
len(nycflight.origin.unique())

3

In [30]:
nycflight.origin.unique()

array(['EWR', 'LGA', 'JFK'], dtype=object)

* There are three NYC airports included in this data. They are EWR, LGA, and JFK.

2. (4pt) Into how many airports did the airlines fly from NYC in 2013?

In [31]:
# check if all dataset is in 2013 
(nycflight.year == 2013).sum()

336776

In [32]:
len(nycflight.dest.unique())

105

* There were 105 airports (or destination) where the airlines fly from NYC in 2013. 

3. (5pt) How many flights were there from NYC to Seattle (airport code SEA)?

In [33]:
(nycflight.dest == "SEA").sum()

3923

* There were 3923 flights from NYC to Seattle in 2013. 

4. (5pt) Were there any flights from NYC to Spokane (GEG)?

In [34]:
(nycflight.dest == "GEG").sum()

0

* Since the sum function returns 0, there were no flight from NYC to Spokane. 

### 2.3 Flight Delays

1. (4pt) Create a new data frame, a subset of the flights data that only contains two variables: departure
delay and arrival delay.

In [35]:
delay = nycflight[["dep_delay", "arr_delay"]]

In [36]:
# check if created dataframe correctly
delay.head(5)

dep_delay  arr_delay
0        2.0       11.0
1        4.0       20.0
2        2.0       33.0
3       -1.0      -18.0
4       -6.0      -25.0

2. (2pt) Print a small random sample of the delay data frame.

In [37]:
delay.sample(10)

dep_delay  arr_delay
139635       -4.0       16.0
181758       -5.0      -17.0
115800       -5.0      -10.0
277224        0.0       13.0
29080        -8.0      -22.0
27129        -6.0      -28.0
232281       -4.0      -44.0
315430       -9.0      -23.0
45320        15.0       27.0
36334         3.0       -9.0

* Above table shows ten random sample from flight delay. 

3. (3pt) What is the data type of the flight delay value?

In [38]:
delay.dtypes

dep_delay    float64
arr_delay    float64
dtype: object

* The data type of both departure delay and arrival delay is float64. 

4. (5pt) Filter to keep the “slow flights”: cases where the plane leaves early but arrives late. How
many cases do you have?
Hint: about 35k

In [39]:
slow_flights = ((delay.dep_delay < 0) & (delay.arr_delay > 0)).sum()
slow_flights

35648

* There are 35648 cases of slow flights: plane leaves early but arrives late. 

5. (5pt) What is the longest “extra flight time”, i.e. difference between arrival delay and departure
delay?
Hint: you can create a new variable that is the extra delay, and find the max value of it.
Hint2: read python notes 5.2 What are the values?

In [40]:
extra_delay = (delay.arr_delay - delay.dep_delay)
# check if extra_delay correctly created
extra_delay.sample(10)

235330     NaN
28044    -22.0
125127   -37.0
322178   -20.0
121927   -30.0
97251      2.0
35068    -27.0
100609   -26.0
27376    -11.0
170507   -23.0
dtype: float64

In [41]:
extra_delay.max()

196.0

* The longest extra flight time is 196 minutes. 

6. (4pt) Are there any implausible flight delay values?

In [42]:
extra_delay.nlargest(10)

55985     196.0
181270    181.0
287308    165.0
259366    161.0
247336    157.0
270752    154.0
250625    153.0
259407    150.0
270758    150.0
181327    148.0
dtype: float64

In [43]:
extra_delay.isna().sum()

9430

* There seems no implausible flight delay because the ten longest extra flight time are within 210 minutes, which is 3 hours and 30 minutes. However, there are 9430 flight delay where values are not descibed (data recorded as NA). 

7. (5pt) How many of flights were delayed (positive arrival delay)?

In [44]:
(delay.arr_delay > 0).sum()

133004

* There were 133,004 delayed flights. 

8. (5pt) What percentage of flights were delayed more than 15 minutes (at arrival)?
Hint: slightly more than 23%.

In [45]:
((delay.arr_delay > 15).sum()) / (len(delay.arr_delay))

0.2305092999501152

* 23.05 percent of flight were delayed more than 15 minutes. 